### PR TITLE
BUG: pd.concat does not drop DatetimeIndex.freq when result is monotonic with even spacing

### LINKED ISSUE #3232: pd.concat loses frequency attribute for 'continuous' DataFrame appends
Hey all, 

I have a `DataFrame` (df) that stores live sensor data that is captured at a specific frequency. New raw data from sensor is updated at a set interval (an attempt at bandwidth conservation),  which is parsed into a new df object.

These new update dataframes are of the same frequency, and contain data that is 'continuous' in time (i.e., they pick up right where the last timestamp left off), and ultimately I would like to append this new data to the existing dataframe while preserving the main dataframe frequency attribute. I tried by using a `concat` of old and new dataframes, however, it seems that `concat` doesn't check this case for continuous time series, and loses its frequency attribute. This can be reproduced in code below:

``` python
import pandas as pd
import numpy as np
dr = pd.date_range('01-Jan-2013', periods=100, freq='50L', tz='UTC')
df = pd.DataFrame(np.random.randn(100, 2), index=dr)
df.index
```

``` python
<class 'pandas.tseries.index.DatetimeIndex'>
[2013-01-01 00:00:00, ..., 2013-01-01 00:00:04.950000]
Length: 100, Freq: 50L, Timezone: UTC
```

These guys look good:

``` python
#Preserves frequency
print df[:50].index
print df[50:].index
```

``` python
<class 'pandas.tseries.index.DatetimeIndex'>
[2013-01-01 00:00:00, ..., 2013-01-01 00:00:02.450000]
Length: 50, Freq: 50L, Timezone: UTC
<class 'pandas.tseries.index.DatetimeIndex'>
[2013-01-01 00:00:02.500000, ..., 2013-01-01 00:00:04.950000]
Length: 50, Freq: 50L, Timezone: UTC
```

However, these guys, together, forget where they came from:

``` python
#Loses frequency
pd.concat([df[:50], df[50:]]).index
```

``` python
<class 'pandas.tseries.index.DatetimeIndex'>
[2013-01-01 00:00:00, ..., 2013-01-01 00:00:04.950000]
Length: 100, Freq: None, Timezone: UTC
```

I currently get around this with a `resample` of the resulting df to set frequency, which isn't that big of a deal, however, thought I'd mention it so that a more elegant behavior could be implemented. I'll try and take a look when I have time, but I know that all you here are so much more familiar with pandas internals. Any pointers?
And, as always, thank you! :)

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #3232
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/indexes/datetimelike.py [hunk 1]
```diff
@@ -587,11 +587,15 @@ def _concat_same_dtype(self, to_concat, name):
         if len({str(x.dtype) for x in to_concat}) != 1:
             raise ValueError('to_concat must have the same tz')
 
-        if not is_period_dtype(self):
+        new_data = type(self._values)._concat_same_type(to_concat).asi8
+
+        # GH 3232: If the concat result is evenly spaced, we can retain the
+        # original frequency
+        is_diff_evenly_spaced = len(np.unique(np.diff(new_data))) == 1
```
Comment (jreback) [inline pandas/core/indexes/datetimelike.py]: I *think* we have something that does this in tseries.frequencies, can you see / use it?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/datetimelike.py (hunk 1 above)
Comment (mroeschke) [reply pandas/core/indexes/datetimelike.py]: Checked in `tseries.frequencies` but didn't find anything. We have `is_monotonic` functions but it doesn't check for even spacing.

Comment (jreback) [reply pandas/core/indexes/datetimelike.py]: you can use unique_1d here (should be faster), though not a big deal. worth moving this actually to tseries.frequencies as a named function?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/datetimelike.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: this looks a bit like `unique_deltas`

Comment (mroeschke) [reply pandas/core/indexes/datetimelike.py]: Thanks @jbrockmendel. Exactly what I needed.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. merge on green.

**Comment by jreback** [comment]
thanks @mroeschke

**Pull request merged**